virttest/qemu_monitor.py

```diff
@@ -598,6 +598,23 @@ def close(self):
         finally:
             self._log_lock.release()
 
+    def wait_for_migrate_progress(self, target):
+        """
+        Wait for migration progress to hit a target %
+        Note: We exit if we've gone onto another pass rather than wait
+        for a target we might never hit.
+        """
+        old_progress = 0
+        while True:
+            progress = self.get_migrate_progress()
+            if (progress < old_progress or
+                    progress >= target):
+                break
+            # progress < old_progress indicates we must be on
+            # another pass (we could also check the sync count)
+            old_progress = progress
+            time.sleep(0.1)
```

Comment:
Yep, this could be useful, although it does not cover the problem I mentioned. The test-facing APIs should be very simple to use in order to allow test writers to simply enable the new feature. Ideally it should not require any changes to tests and it should simply work out of the box.

Take external migration for example. It was added and the only change required to tests is to add:

```diff
                 vm.migrate(mig_timeout, mig_protocol, mig_cancel_delay,
-                           env=env)
+                           env=env,
+                           migration_exec_cmd_src=migration_exec_cmd_src,
+                           migration_exec_cmd_dst=migration_exec_cmd_dst)
```

With that said, would you consider adding some default values for `mig_inner_funcs`? I'd expect something like:

```python
# this should be private function inside `def migrate`
def _inner_func(func)
    func = func.split(None, 1)
    if len(func) > 1:
        func, arg = func
    else:
        func = func[0]
        arg = None
    # Default inner_funcs goes here
    if func == "postcopy"
        if arg is None:
            arg = 70
        vm.monitor.wait_for_migrate_progress(arg)
        vm.monitor.migrate_start_postcopy()

...

if mig_inner_funcs:
    for func in mig_inner_funcs:
        if isinstance(func, six.string_types):
            _inner_func(func)
        else:
            func()
```

Reply:
PS: I know the `mig_inner_func` is free-form, super flexible, but in reality I'd only use the built-in function unless it was single-purpose test tailored to check this function. And in such case I'd probably just use `not_wait_for_migration` and do the special handling there.

Reply:
OK, just posted a new version with a specified inner function; let me know what you think